### PR TITLE
PP-5358 Add sandbox mandate state calculator

### PR DESCRIPTION
**Pull request by stephencdaly**
Add state calculator for sandbox mandates that only takes into account GovUkPayEvents as no SandboxEvents affect the state. Rename GoCardlessMandateStateUpdater to MandateStateUpdater and make it use the appropriate calculator depending on the PaymentProvider of the gateway account for the mandate it is calculating for.

### CODE REVIEW

File: src/main/java/uk/gov/pay/directdebit/mandate/services/GovUkPayEventToMandateStateMapper.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import static uk.gov.pay.directdebit.events.model.GovUkPayEventType.MANDATE_TOKEN_EXCHANGED;
+import static uk.gov.pay.directdebit.mandate.model.MandateState.AWAITING_DIRECT_DEBIT_DETAILS;
+import static uk.gov.pay.directdebit.mandate.model.MandateState.CANCELLED;
+import static uk.gov.pay.directdebit.mandate.model.MandateState.CREATED;
+import static uk.gov.pay.directdebit.mandate.model.MandateState.EXPIRED;
+import static uk.gov.pay.directdebit.mandate.model.MandateState.SUBMITTED;
+import static uk.gov.pay.directdebit.mandate.model.MandateState.USER_CANCEL_NOT_ELIGIBLE;
+
+public class GovUkPayEventToMandateStateMapper {
+    private static final Map<GovUkPayEventType, MandateState> GOV_UK_PAY_EVENT_TYPE_TO_MANDATE_STATE = Map.of(
```
Comment (alexbishop1) [inline src/main/java/uk/gov/pay/directdebit/mandate/services/GovUkPayEventToMandateStateMapper.java]: Rare opportunity for an `EnumMap` here but we’d lose guaranteed unmodifiability and the ease of initialisation so maybe it’s not worth it

File: src/main/java/uk/gov/pay/directdebit/mandate/services/GovUkPayEventToMandateStateMapper.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+            MANDATE_EXPIRED_BY_SYSTEM, EXPIRED,
+            MANDATE_CANCELLED_BY_USER, CANCELLED,
+            MANDATE_CANCELLED_BY_USER_NOT_ELIGIBLE, USER_CANCEL_NOT_ELIGIBLE
+    );
+
+    public static final Set<GovUkPayEventType> GOV_UK_PAY_EVENT_TYPES_THAT_CHANGE_STATE = GOV_UK_PAY_EVENT_TYPE_TO_MANDATE_STATE.keySet();
+
+    public static Optional<DirectDebitStateWithDetails<MandateState>> mapGovUkPayEventToState(GovUkPayEvent govUkPayEvent) {
+        return Optional.of(GOV_UK_PAY_EVENT_TYPE_TO_MANDATE_STATE.get(govUkPayEvent.getEventType()))
+                .map(DirectDebitStateWithDetails::new);
```
Comment (alexbishop1) [inline src/main/java/uk/gov/pay/directdebit/mandate/services/GovUkPayEventToMandateStateMapper.java]: Did you mean to use `Optional.ofNullable(…)` here?

File: src/main/java/uk/gov/pay/directdebit/mandate/services/MandateStateUpdater.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+                .calculate(mandate)
+                .ifPresentOrElse(stateAndDetails -> mandateUpdateService.updateState(mandate, stateAndDetails),
+                        () -> LOGGER.info(format("Asked to update the status for mandate %s but there appear to be " +
+                                "no events stored that require it to be updated", mandate.getExternalId())));
+    }
+
+    private MandateStateCalculator getStateCalculator(Mandate mandate) {
+        return mandate.getGatewayAccount().getPaymentProvider() == GOCARDLESS
+                ? goCardlessMandateStateCalculator
+                : sandboxMandateStateCalculator;
```
Comment (alexbishop1) [inline src/main/java/uk/gov/pay/directdebit/mandate/services/MandateStateUpdater.java]: Maybe use a `switch` statement here?

File: src/test/java/uk/gov/pay/directdebit/mandate/services/MandateStateUpdaterTest.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    public void updatesMandateWithStateReturnedByCalculatorForGoCardlessMandate() {
+        GatewayAccountFixture gatewayAccountFixture = aGatewayAccountFixture().withPaymentProvider(GOCARDLESS);
+        Mandate mandate = aMandateFixture().withGatewayAccountFixture(gatewayAccountFixture).toEntity();
+
+        given(mockGoCardlessMandateStateCalculator.calculate(mandate))
+                .willReturn(Optional.of(mockMandateStateWithDetails));
+
+        mockMandateStateUpdater.updateState(mandate);
+
+        verify(mockGoCardlessMandateStateCalculator).calculate(mandate);
```
Comment (alexbishop1) [inline src/test/java/uk/gov/pay/directdebit/mandate/services/MandateStateUpdaterTest.java]: No need to verify something we’re already stubbing

File: src/test/java/uk/gov/pay/directdebit/mandate/services/MandateStateUpdaterTest.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        verify(mockMandateUpdateService).updateState(mandate,
+                mockMandateStateWithDetails);
```
Comment (alexbishop1) [inline src/test/java/uk/gov/pay/directdebit/mandate/services/MandateStateUpdaterTest.java]: These lines don’t need to be this short: it’s not 1985 any more

File: src/main/java/uk/gov/pay/directdebit/mandate/services/GovUkPayEventToMandateStateMapper.java (hunk 2 above)
Comment (stephencdaly) [reply src/main/java/uk/gov/pay/directdebit/mandate/services/GovUkPayEventToMandateStateMapper.java]: yes i did, moving around the code a bit between branches made me lose that

File: src/main/java/uk/gov/pay/directdebit/mandate/services/MandateStateUpdater.java (hunk 3 above)
Comment (stephencdaly) [reply src/main/java/uk/gov/pay/directdebit/mandate/services/MandateStateUpdater.java]: ok

**Comment by stephencdaly** [comment]
Added requested changes in a new commit

**Pull request merged**